### ISSUE TITLE
[bug] Conan mixed requires which have different options

### ISSUE THREAD

**Issue by zhanglin-zhou** [conan-io/conan#15865]
### Environment details

* Operating System+version: Windows 10
* Compiler+version: msvc 192
* Conan version: 2.0.17
* Python version:  3.11.5

### Steps to reproduce

1. My conan file:
```
from conan import ConanFile

class ConanTest(ConanFile):
    settings = "os", "arch", "compiler", "build_type"

    def requirements(self):
        self.requires("libpng/1.6.43")
        self.requires("zlib/1.3.1", options={"shared": True})
```
2. My build profile:
```
[settings]
os=Windows
arch=x86_64
```
3. My host profile
```
[settings]
os=Windows
arch=x86_64
compiler=msvc
compiler.version=192
compiler.runtime=dynamic
build_type=Release
```
4.  Run following command:
```
conan install test.py -pr:b build.profile -pr:h host.profile -r conancenter
```
5. Only zlib with shared=False is installed:
```
conan list zlib*#*:*

Local Cache
  zlib
    zlib/1.3.1
      revisions
        f52e03ae3d251dec704634230cd806a2 (2024-02-22 09:20:06 UTC)
          packages
            41ad450120fdab2266b1185a967d298f7ae52595
              info
                settings
                  arch: x86_64
                  build_type: Release
                  compiler: msvc
                  compiler.runtime: dynamic
                  compiler.runtime_type: Release
                  compiler.version: 192
                  os: Windows
                options
                  shared: False
```
The reason is conan doesn't force options to be equal while checking the equality of require here:
https://github.com/conan-io/conan/blob/bdefa4df7aa3888f87aebbe2f1db1384c6a448be/conans/model/requires.py#L218

so the zlib with options `shared=True` in my conanfile is overridden by libpng's zlib dependency, which has options `shared=False`.

**Comment by memsharded** [comment]
Hi @zhanglin-zhou 

Thanks for your feedback.

It seems this is expected behavior. Conan follows a depth-first graph expansion, so dependencies that expand first will have higher priority.

In this case you have:
```python
 def requirements(self):
        self.requires("libpng/1.6.43")
        self.requires("zlib/1.3.1", options={"shared": True})
```
That means that ``libpng`` expansion will have higher priority, and if ``libpng`` by default is requiring ``zlib`` as static library, that will still have precedence. You can change the priority by:
```python
 def requirements(self):
        self.requires("zlib/1.3.1", options={"shared": True})
        self.requires("libpng/1.6.43")
```
Or by:
```python
 def requirements(self):
        self.requires("libpng/1.6.43", options={"zlib*:shared": True})
        self.requires("zlib/1.3.1", options={"shared": True})
```

In general, it is better recommended to define variability via ``options`` in the ``profiles`` files, not in recipes, as it guarantees the global definition of the options values.

**Comment by zhanglin-zhou** [comment]
I can understand what happen here, but I have concern about this behavior, because it doesn't warn me or throw error while this happen, zlib/1.3.1 with shared=True and zlib/1.3.1 with shared=False are two different packages, I have explicitly stated that I want a shared zlib in my conanfile, and it's replaced with a static zlib without any warning. I am lucky here because my product build process throws error while linking a static zlib, so I can note this issue. 
But considering other shared library with multiple options like openssl, how can I know which openssl will be used? Ok, the depth-first one, then where is the depth-first one come from? I need to go through all the packages in my conan file and its dependency to find when the openssl is first introduced, then check whether its options can satisfy my product's requirement.
This is really really crazy while building a complex product.
Actually I think while comparing two requires, the options must be compared too, because it does affect the package binary's behavior, and options belong to package itself, what do you think?

**Comment by memsharded** [comment]
> I can understand what happen here, but I have concern about this behavior, because it doesn't warn me or throw error while this happen, zlib/1.3.1 with shared=True and zlib/1.3.1 with shared=False are two different packages, I have explicitly stated that I want a shared zlib in my conanfile, and it's replaced with a static zlib without any warning. I am lucky here because my product build process throws error while linking a static zlib, so I can note this issue.

The recommended way to always avoid using certain dependencies options is to implement a ``validate()`` method that raises if the upstream options aren't valid for the current package. Otherwise the current recipe ``options`` values definitions for dependencies are never enforced, it is just a low priority recommendation that will be followed if not overriding by downstream, and this recipe has higher priority than others (being required first).

> But considering other shared library with multiple options like openssl, how can I know which openssl will be used? Ok, the depth-first one, then where is the depth-first one come from? I need to go through all the packages in my conan file and its dependency to find when the openssl is first introduced, then check whether its options can satisfy my product's requirement.
This is really really crazy while building a complex product.

Recipes do not define options, recipes can only suggest options. The recommended way to guarantee some options values is defining them in the ``profile`` files, not in recipes.

> Actually I think while comparing two requires, the options must be compared too, because it does affect the package binary's behavior, and options belong to package itself, what do you think?

No, I am afraid that is not possible. The comparison of requires is designed to detect dependencies conflicts. Having different options will still cause conflicts, even if the requirements are considered different, because you cannot link 2 different variants of the same library even if they are built with different options. This is the reason why the comparison of requires is mainly based on the ``traits`` that cause compilation and linkage errors in C and C++, like headers, libraries or runtime libraries. Having 2 different dependencies of ``zlib`` just because one define ``with_minizip=True`` and the other doesn't is an error in the vast majority of cases.

**Comment by zhanglin-zhou** [comment]
> The recommended way to always avoid using certain dependencies options is to implement a `validate()` method that raises if the upstream options aren't valid for the current package. Otherwise the current recipe `options` values definitions for dependencies are never enforced, it is just a low priority recommendation that will be followed if not overriding by downstream, and this recipe has higher priority than others (being required first).

Yeah, I understand this behavior after several hours debugging. My point is, if you look back my conan file:
```
    def requirements(self):
        self.requires("libpng/1.6.43")
        self.requires("zlib/1.3.1", options={"shared": True})
```
and I get a static zlib finally, don't you think it's counterintuitive?

> Recipes do not define options, recipes can only suggest options. The recommended way to guarantee some options values is defining them in the `profile` files, not in recipes.

> because you cannot link 2 different variants of the same library even if they are built with different options. 

I guess you mean we cannot directly link 2 different variants of the same library into one executable file, right? Because there is case that an executable file links zlib dynamically, and then links libpng which links zlib statically. Also in our case, the product have multiple executable files, so different executable may requires different options.

Also my point is not about avoiding conflict, my point is, conan should warn me or give me error when this kind of conflict happen, instead of replacing the package silently.

**Comment by memsharded** [comment]
> Also my point is not about avoiding conflict, my point is, conan should warn me or give me error when this kind of conflict happen, instead of replacing the package silently.

It might be possible for Conan to try to give some warning for this specific pattern with ``{"shared": True}``.
However, not that it is enough for the downstream consumer of this recipe to do ``{zlib/*:shared: True}`` and then even if you have that in your recipe, it will not even warn, because the ``self.requires(..., options=..`` will be overridden by the downstream even before checking the previous values of ``zlib`` options.

Let's see if it is possible to give that warning for this local use case.

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/15876 added some information about mismatches in options values, will be included in Conan 2.2
